content/blog/2019-10-3-tdd-gimmick/index.md

```diff
@@ -0,0 +1,196 @@
+- Go back over your implementation and clean-up.
+  - This is the `REFACTOR` phase.
+
+In practice it might look like this. We'll take a simple example to start with.
+
+Say, I want to write a function `mul2` that returns the product of two arguments:
+
+First create the test-file: `mul2.test.ts` and start writing the test:
+
+```ts
+it('should return the product', () => {
+  const actual = mul2()
+})
+```
+
+As soon as this is written, the compiler will complain that `mul2` doesn't exist. Only at this point do you actually create the `mul2.ts` file and write:
+
+```ts
+export const mul2 = () => {
+  throw 'not implemented'
+}
+```
+
+All you know is that it's a function, which is what you've declared. No implementation yet, no arguments either.
+
+Back to the test:
+
+```ts
+it('should return the product', () => {
+  const actual = mul2(2, 3)
+})
+```
+
+The compiler will complain: "Function doesn't accept values". Fair enough, now is a good time to think about the "API" of our function:
+
+- What are the order of the arguments,
+  - important if the function might be curried
+- Can any of them be null? do any of them have default values? What are their names?
+- What's the return type of the function?
+
+```ts
+export const mul2 = (a: number, b: number) => {
+  throw 'not implemented'
+}
+```
+
+Since it's just a simple function, the types are pretty basic but if there were more complex behaviour, here's where we would take care of that[^typings].
+
+Back to the test, now we can add our assertion:
+
+```ts
+it('should return the product', () => {
+  const actual = mul2(2, 3)
+  const expected = 6
+  assert.strictEqual(actual, expected)
+})
+```
+
+The test itself is now runnable. Run it, and it'll throw "Not Implemented" as we told it to. This is the point people skip to when they think of unit-tests, but it takes a lot of thinking even to get to this point; we're just so used to doing it we forget to account for it.
+
+```ts
+export const mul2 = (a: number, b: number) => {
+  return a * b
+}
+```
+
+With that your test will pass, and you can start with the next test!
+
+## But this way of writing tests must be super annoying
+
+Writing code this way is definitely a little unnatural at first, but it makes sure that nothing extraneous ever enters your code base. When people talk about Lean-Testing, and only writing the minimum tests required, this is really what that looks like; no more tests than necessary, no more code than needed.
+
+Over time the intuition for this style will build and your compiler will be the best pair-programming partner you could ever ask for.
+
+## Cool, but I could just write the function and then write the test afterwards
+
+As far as I've spoken to people, no one disagrees with the need to have some unit tests. That means we all agree that it's possible to write buggy code. That implies that our unit-tests could _also_ have bugs; how would we detect that if we wrote it second?
```

Comment:
I usually get the complain "if I write the test first, instead of after, I will need to keep updating the tests until I finish the implementation".

Reply:
this is a good one, I've tried to add an answer.